helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -225,10 +227,52 @@ protected StatefulRebalancer createRebalancer(HelixManager helixManager) {
    *  1) ZK callback should go to ClusterDataCache and trigger data cache refresh only
    *  2) then ClusterDataCache.refresh triggers rebalance pipeline.
    */
-  /* Map of cluster->GenrichelixController */
-  private static Map<String, GenericHelixController> HelixControllerFactory = new ConcurrentHashMap<>();
-  public static GenericHelixController getController(String clusterName) {
-    return HelixControllerFactory.get(clusterName);
+  /* Map of cluster->Set of GenericHelixController*/
+  private static Map<String, ImmutableSet<GenericHelixController>> _helixControllerFactory =
+      new ConcurrentHashMap<>();
+
+  public static GenericHelixController getLeaderController(String clusterName) {
```

Comment:
Question: Is this necessary? This is usually in memory triggered rebalance. It only triggers when the node is leader for the cluster. Also you should not see one generic helix controller perform as LEADER and STANDBY at same time for a cluster. Do we need this set?

Reply:
This problem fails more than 2 tests. In production, and if the application dev is reasonable, then it won't appear. 
My take is that 1. the new change is harmless for the production logic. 2. if not doing this, we have to hack our test logic. So it may hide some real problem. Since the test is not doing what production code does.
So it is really necessary.

Reply:
Same question, why we could have two controllers for same cluster? Is this just happening in tests because the controller from other tests were not properly closed?

Reply:
There are more than one test cases need to start multiple controllers. If not doing so, then all these tests have to trigger the rebalance manually, since the auto scheduled ondemand rebalance won't work. This is a potential problem since the workaround hide problems.